src/python/WMCore/ReqMgr/CherryPyThreads/StepChainParentageFixTask.py

```diff
@@ -51,52 +47,57 @@ def fixStepChainParentage(self, config):
         Fix the StepChain parentage and update the ParentageResolved flag to True
         """
         self.logger.info("Running fixStepChainParentage thread for statuses: %s", self.statusToCheck)
-        childDatasets = set()
-        requests = set()
-        requestsByChildDataset = {}
+        fixedCount = 0
         for status in self.statusToCheck:
-            reqByChildDS = getChildDatasetsForStepChainMissingParent(self.reqmgrDB, status)
-            self.logger.info("Retrieved %d datasets to fix parentage, in status: %s",
-                             len(reqByChildDS), status)
-            childDatasets = childDatasets.union(set(reqByChildDS.keys()))
-            # We need to just get one of the StepChain workflow if multiple workflow contains the same datasets. (i.e. ACDC)
-            requestsByChildDataset.update(reqByChildDS)
-
-            for wfs in viewvalues(reqByChildDS):
-                requests = requests.union(wfs)
-
-        failedRequests = set()
-        totalChildDS = len(childDatasets)
-        fixCount = 0
-        for childDS in childDatasets:
-            self.logger.info("Resolving parentage for dataset: %s", childDS)
-            start = time.time()
-            try:
-                failedBlocks = self.dbsSvc.fixMissingParentageDatasets(childDS, insertFlag=True)
-            except Exception as exc:
-                self.logger.exception("Failed to resolve parentage data for dataset: %s. Error: %s",
-                                      childDS, str(exc))
-                failedRequests = failedRequests.union(requestsByChildDataset[childDS])
-            else:
-                if failedBlocks:
-                    self.logger.warning("These blocks failed to be resolved and will be retried later: %s",
-                                        failedBlocks)
-                    failedRequests = failedRequests.union(requestsByChildDataset[childDS])
+            listWflowChildren = getChildDatasetsForStepChainMissingParent(self.reqmgrDB, status)
+            msg = f"Retrieved {len(listWflowChildren)} workflows to fix parentage in status {status}"
+            self.logger.info(msg)
+
+            # now resolve every dataset parentage for each workflow
+            while listWflowChildren:
+                wflowData = listWflowChildren.pop()
```

Comment:
The `pop` method can throw exception if it is empty, e.g.
```
>>> d=[]
>>> d.pop()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: pop from empty list
```
And, code here does not check for it. It is better to put it into try/except block.

Reply:
This is what the `while listWflowChildren:` is supposed to do. E.g.:
```
In [1]: l = [1, 2, 3]

In [2]: while l:
   ...:     a = l.pop()
   ...:     print(a)
   ...: 
3
2
1
```